examples/so_vits_svc.py

```diff
@@ -0,0 +1,834 @@
+    self.num_kernels, self.num_upsamples = len(resblock_kernel_sizes), len(upsample_rates)
+    self.conv_pre = nn.Conv1d(inter_channels, upsample_initial_channel, 7, 1, padding=3)
+    self.f0_upsamp = Upsample(scale_factor=np.prod(upsample_rates))
+    self.m_source = SourceHnNSF(sampling_rate, harmonic_num=8)
+    resblock = ResBlock1 if resblock == '1' else ResBlock2
+    self.ups, self.noise_convs = [], []
+    for i, (u, k) in enumerate(zip(upsample_rates, upsample_kernel_sizes)):
+      c_cur = upsample_initial_channel//(2**(i+1))
+      self.ups.append(nn.ConvTranspose1d(upsample_initial_channel//(2**i), c_cur, k, u, padding=(k-u)//2))
+      if i + 1 < len(upsample_rates):  # TODO: make oneliner after debugging
+        stride_f0 = int(np.prod(upsample_rates[i + 1:]))
+        self.noise_convs.append(nn.Conv1d(1, c_cur, kernel_size=stride_f0 * 2, stride=stride_f0, padding=(stride_f0+1) // 2))
+      else:
+        self.noise_convs.append(nn.Conv1d(1, c_cur, kernel_size=1))
+    self.resblocks = []
+    for i in range(len(self.ups)):
+      ch = upsample_initial_channel // (2 ** (i + 1))
+      for _, (k, d) in enumerate(zip(resblock_kernel_sizes, resblock_dilation_sizes)):
+        self.resblocks.append(resblock(ch, k, d))
+    self.conv_post = nn.Conv1d(ch, 1, 7, 1, padding=3)
+    if gin_channels != 0: self.cond = nn.Conv1d(gin_channels, upsample_initial_channel, 1)
+    self.upp = np.prod(upsample_rates)
+  def forward(self, x, f0, g=None):
+    f0 = self.f0_upsamp.forward(f0[:, None]).transpose(1, 2)  # bs,n,t
+    har_source, _, _ = self.m_source.forward(f0, self.upp)
+    har_source = har_source.transpose(1, 2)
+    x = self.conv_pre(x)
+    if g is not None:  x = x + self.cond(g)
+    for i in range(self.num_upsamples):
+      x, xs = self.ups[i](x.leakyrelu(LRELU_SLOPE)), None
+      x_source = self.noise_convs[i](har_source)
+      x = x + x_source
+      for j in range(self.num_kernels):
+        if xs is None: xs = self.resblocks[i * self.num_kernels + j].forward(x)
+        else: xs += self.resblocks[i * self.num_kernels + j].forward(x)
+      x = xs / self.num_kernels
+    return self.conv_post(x.leakyrelu()).tanh()
+
+def repeat_expand_2d(content, target_len, mode="left"): # content : [h, t]
+  if mode == "left": return repeat_expand_2d_left(content, target_len)
+  elif mode == "nearest": return repeat_expand_2d_nearest(content, target_len)
+  raise RuntimeError("Mode: {mode} not supported.")
+
+def repeat_expand_2d_left(content, target_len): # content : [h, t]
+  content = content.numpy()
+  target = np.zeros((content.shape[0], target_len), dtype=np.float32)
+  src_len = content.shape[-1]
+  target = np.zeros((content.shape[0], target_len), dtype=np.float32)
+  temp = np.arange(src_len+1) * target_len / src_len
+  current_pos = 0
+  for i in range(target_len):
+    if i >= temp[current_pos+1]:
+      current_pos += 1
+    target[:, i] = content[:, current_pos]
+  return Tensor(target)
```

Comment:
Can we do something with `Tensor.stack` or `Tensor.cat` here? Just trying to get rid of the numpy usage as much as possible.

Reply:
[7646492](https://github.com/tinygrad/tinygrad/pull/1371/commits/76464921fbf3727893a62191b7b2674487dadb23)